### PR TITLE
Double buffering for Hopper arch

### PR DESCRIPTION
**Pull request by drzejan2**
The items of this change:
- prepare new set of double buffer test cases, targeting H100 arch
- update implementation of indecing computation for cpAsyncBulk operations

### CODE REVIEW

**Comment by drzejan2** [comment]
I changed the approach for adding the support Hopper TMA double buffering / circular buffering - I decided to focus on how the final kernel should look like.

Note that code is based on currently generated double buffering so any variable starting with `a` is a result of manually adding variable to the sources.

The generated kernel is a result of running the following test case ([link](https://github.com/NVIDIA/Fuser/pull/1484/files#diff-b5f10578a9754f9333df9e53cf1e74c4cf3d3cb2c20998de8e740269dc03debeR38)).

@naoyam can I ask you to check if prepared kernels make sense? I will start working on preparing changes in the infrastructure to produce code like below.

@zasdfgbnm for visibility.

----

<details>
<summary>Double buffer kernel</summary>

```
__global__ void nvfuser_none_f0_c0_r0_g0(Tensor<float, 1, 1> T0, const __grid_constant__ TensorMap var0, Tensor<float, 1, 1> T2) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  NVFUSER_DEFINE_MAGIC_ZERO;
  nvfuser_index_t i1;
  i1 = ceilDiv(T0.logical_size[0LL], 32);
  const TensorMap* ptr2;
  ptr2 = &var0;

  // NOTE: the offset between mbarrier objects in an array of mbarrier objects
  // NOTE: barrier_t is not defined in this scope, can we include barrier.h?
  unsigned a10;
  a10 = (unsigned)sizeof(barrier_t);

  // NOTE: the number of expected 'arrivals' of mbarrier, used for barrier init
  uint64_t a11;
  a11 = blockDim.x * blockDim.y * blockDim.z;

  // NOTE: placeholder for tokens
  uint64_t a12[<stages>];

  // NOTE: buffer for loaded gmem, size of 128 floats, see T3 definition
  float* T1 = reinterpret_cast<float*>(array + smem_offset + 0);
  
  // NOTE: start address of a placeholder for mbarrier objects
  uint64_t* T3 = reinterpret_cast<uint64_t*>(array + smem_offset + (128LL * <stages>));
  
  // NOTE: init all mbarrier objects by the first thread in the block
  if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
    for (unsigned a4 = 0; a4 < 2; ++a4) {
      mbarrier::init((toSmem(T3) + a10 * a4), a11);
    }
  }
  
  // NOTE: unrolled prolog loop
  // NOTE: can be merged with if expr with mbarrier init
  if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
    a12[0] = mbarrier::arriveExpectTX((toSmem(T3) + a10 * 0), (uint32_t)((((ceilDiv(T0.logical_size[0LL], 32)) * 4LL) * 32LL)));
    Hopper::cpAsyncBulkTensorTileG2S((Hopper::CpAsyncBulkTensorTileG2SIndex<1>{ ptr2, (Array<nvfuser_index_t, 1, 1>{(32 * 0)}), (toSmem(T3) + a10 * 0) }), (toSmem(T1) + (128 * (0));
  }

  // NOTE: wait for all threads in block
  block_sync::sync<0>();

  // NOTE: main loop
  NVFUSER_UPDATE_MAGIC_ZERO;
  #pragma unroll 1
  for(nvfuser_index_t i6 = 0; i6 < (i1 - 1); ++i6) {
    nvfuser_index_t i7;
    i7 = 32LL * (i6 % 2);
    nvfuser_index_t i8;
    i8 = 32LL * i6;

    // NOTE: index of the current barrier
    nvfuser_index_t a20;
    a20 = (i6 % 2);

    // NOTE: index of the next stage to fetch
    nvfuser_index_t a21;
    a21 = ((i6 + 1) % 2);

    // NOTE: schedule the next stage in the queue - a21
    if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
      a12[a21] = mbarrier::arriveExpectTX((toSmem(T3) + a10 * a21), (uint32_t)((((ceilDiv(T0.logical_size[0LL], 32)) * 4LL) * 32LL)));
      Hopper::cpAsyncBulkTensorTileG2S((Hopper::CpAsyncBulkTensorTileG2SIndex<1>{ ptr2, (Array<nvfuser_index_t, 1, 1>{(32 * i6)}), (toSmem(T3) + a10 * a21) }), (toSmem(T1) + (128 * a21)));
    }

    // NOTE: wait for the current barrier - a20
    // NOTE: 1st: mbarrier, 2nd: the expected token for barrier
    mbarrier::wait((toSmem(T3) + a10 * a20), a12[a20]);

    NVFUSER_UPDATE_MAGIC_ZERO;
    #pragma unroll
    for(nvfuser_index_t i10 = 0; i10 < 32; ++i10) {
      nvfuser_index_t i11;
      i11 = i8 + (i10 + nvfuser_zero);
      if ((i11 < T0.logical_size[0LL])) {
        T2[i11]
           = T1[(i7 + i10)];
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  }
  
  // NOTE: epilogue loop
  // NOTE: we can avoid epilogue loop by predicating the fetch cpAsync
  #pragma unroll 1
  for(nvfuser_index_t a30 = (i1 - 1); a30 < i1; ++a30) {
    nvfuser_index_t a31;
    a31 = 32LL * (a30 % 2);
    nvfuser_index_t a32;
    a32 = 32LL * a30;
    
    // NOTE: current barrier index
    nvfuser_index_t a33;
    a33 = (a30 % 2);
    
    // NOTE: wait for the current barrier to be processed - a33
    mbarrier::wait((toSmem(T3) + a10 * a33), a12[a33]);
    
    NVFUSER_UPDATE_MAGIC_ZERO;
    #pragma unroll
    for(nvfuser_index_t i10 = 0; i10 < 32; ++i10) {
      nvfuser_index_t i11;
      i11 = a32 + (i10 + nvfuser_zero);
      if ((i11 < T0.logical_size[0LL])) {
        T2[i11]
           = T1[(a31 + i10)];
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  }
  
  // NOTE: inval memory with mbarrier objects
  if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
    for (unsigned a4 = 0; a4 < <stages>; ++a4) {
      mbarrier::inval((toSmem(T3) + a10 * a4));
    }
  }
}
```
</details>

----

<details>
<summary>Circular buffer kernel</summary>

```
__global__ void nvfuser_none_f0_c0_r0_g0(Tensor<float, 1, 1> T0, const __grid_constant__ TensorMap var0, Tensor<float, 1, 1> T2) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  NVFUSER_DEFINE_MAGIC_ZERO;
  nvfuser_index_t i1;
  i1 = ceilDiv(T0.logical_size[0LL], 32);
  const TensorMap* ptr2;
  ptr2 = &var0;

  // NOTE: the offset between mbarrier objects in an array of mbarrier objects
  // NOTE: barrier_t is not defined in this scope, can we include barrier.h?
  unsigned a10;
  a10 = (unsigned)sizeof(barrier_t);

  // NOTE: the number of expected 'arrivals' of mbarrier, used for barrier init
  uint64_t a11;
  a11 = blockDim.x * blockDim.y * blockDim.z;

  // NOTE: placeholder for tokens
  uint64_t a12[<stages>];

  // NOTE: buffer for loaded gmem, size of 128 floats, see T3 definition
  float* T1 = reinterpret_cast<float*>(array + smem_offset + 0);
  
  // NOTE: start address of a placeholder for mbarrier objects
  uint64_t* T3 = reinterpret_cast<uint64_t*>(array + smem_offset + (128LL * <stages>));
  
  // NOTE: why T3 was reference before used?
  unsigned i3;
  i3 = toSmem(T3);
  
  // NOTE: init all mbarrier objects by the first thread in the block
  if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
    for (unsigned a4 = 0; a4 < <stages>; ++a4) {
      mbarrier::init((toSmem(T3) + a10 * a4), a11);
    }
  }
  
  // NOTE: unrolled prolog loop
  // NOTE: can be merged with if expr with mbarrier init
  if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
    for (unsigned a4 = 0; a4 < (<stages> - 1); ++a4) {
      a12[a4] = mbarrier::arriveExpectTX((toSmem(T3) + a10 * a4), (uint32_t)((((ceilDiv(T0.logical_size[0LL], 32)) * 4LL) * 32LL)));
      Hopper::cpAsyncBulkTensorTileG2S((Hopper::CpAsyncBulkTensorTileG2SIndex<1>{ ptr2, (Array<nvfuser_index_t, 1, 1>{(32 * a4)}), (toSmem(T3) + a10 * a4) }), (toSmem(T1) + (128 * (a4));
    }
  }

  // NOTE: wait for all threads in block
  block_sync::sync<0>();

  // NOTE: main loop
  // NOTE: limit has been changed, 'i1' -> '(i1 - (stage - 1))'
  NVFUSER_UPDATE_MAGIC_ZERO;
  #pragma unroll 1
  for(nvfuser_index_t i6 = 0; i6 < (i1 - (stage - 1)); ++i6) {
    nvfuser_index_t i7;
    i7 = 32LL * (i6 % <stages>);
    nvfuser_index_t i8;
    i8 = 32LL * i6;

    // NOTE: index of the current barrier
    nvfuser_index_t a20;
    a20 = (i6 % <stages>);

    // NOTE: index of the next stage to fetch
    nvfuser_index_t a21;
    a21 = ((i6 + stage - 1) % <stages>);

    // NOTE: schedule the next stage in the queue - a21
    if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
      a12[a21] = mbarrier::arriveExpectTX((toSmem(T3) + a10 * a21), (uint32_t)((((ceilDiv(T0.logical_size[0LL], 32)) * 4LL) * 32LL)));
      Hopper::cpAsyncBulkTensorTileG2S((Hopper::CpAsyncBulkTensorTileG2SIndex<1>{ ptr2, (Array<nvfuser_index_t, 1, 1>{(32 * i6)}), (toSmem(T3) + a10 * a21) }), (toSmem(T1) + (128 * a21)));
    }

    // NOTE: wait for the current barrier - a20
    // NOTE: 1st: mbarrier, 2nd: the expected token for barrier
    mbarrier::wait((toSmem(T3) + a10 * a20), a12[a20]);

    NVFUSER_UPDATE_MAGIC_ZERO;
    #pragma unroll
    for(nvfuser_index_t i10 = 0; i10 < 32; ++i10) {
      nvfuser_index_t i11;
      i11 = i8 + (i10 + nvfuser_zero);
      if ((i11 < T0.logical_size[0LL])) {
        T2[i11]
           = T1[(i7 + i10)];
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  }
  
  // NOTE: epilogue loop
  // NOTE: we can avoid epilogue loop by predicating the fetch cpAsync
  #pragma unroll 1
  for(nvfuser_index_t a30 = (i1 - (<stage> - 1)); a30 < i1; ++a30) {
    nvfuser_index_t a31;
    a31 = 32LL * (a30 % <stages>);
    nvfuser_index_t a32;
    a32 = 32LL * a30;
    
    // NOTE: current barrier index
    nvfuser_index_t a33;
    a33 = (a30 % <stages>);
    
    // NOTE: wait for the current barrier to be processed - a33
    mbarrier::wait((toSmem(T3) + a10 * a33), a12[a33]);
    
    NVFUSER_UPDATE_MAGIC_ZERO;
    #pragma unroll
    for(nvfuser_index_t i10 = 0; i10 < 32; ++i10) {
      nvfuser_index_t i11;
      i11 = a32 + (i10 + nvfuser_zero);
      if ((i11 < T0.logical_size[0LL])) {
        T2[i11]
           = T1[(a31 + i10)];
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  }
  
  // NOTE: inval memory with mbarrier objects
  if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
    for (unsigned a4 = 0; a4 < <stages>; ++a4) {
      mbarrier::inval((toSmem(T3) + a10 * a4));
    }
  }
}
```
</details>

**Comment by drzejan2** [comment]
The state of the implementation, as of 02/22/2024:
- add handling of additional smem allocations needed to store tokens (token - an object used to synchronize memory operations that use mbarrier objects)
- update smem allocations for mbarrier and token objects; double/circular buffering requires a number of stages of instances of these objects, per double-buffered TensorView
- update mechanism for calculating smem allocations' lifetime (e.g. mbarrier starts before its initialize, and ends after invalidation)
- modify indexing - we shift focus from allocation/index pass to double buffering pass w.r.t. cpAsyncBulk memory operations, but there are still cases where memory operation is cpAsyncBulk, but it is not a part of double buffering - for these cases we keep naïve synchronization mechanism, but at the same time we don't want to break what was done in Double Buffering pass (`LoadStoreOp` replacements / new loop introduction)
- modify Double Buffering pass in lowering
  - separate execution paths, based on `LoadStoreOp`s in Double Buffer loop:
    - if there is no `LoadStoreOp` that is cpAsyncBulk, then use original (but with some changes) execution path (see `insert`)
    - if there is at least one `LoadStoreOp` that is cpAsyncBulk, then use `insertTMA` (name to be changed) which redefines structure of result kir
  - new execution path introduces two new concepts, apart form existing prolog/main/epilogue loops (more details can be found at the end of this long comment):
    - 'pre-prolog' - a collection of expressions that contain smem allocations (so these are visible through the double-buffered loop) and `mbarrier` initialization, which is predicated so it is done only by the first thread in a block,
    - `post-epilogue` - a collection of expressions that handle `mbarrier` objects invalidation,
  - to handle `cpAsyncBulk` / `mbarrier::wait` in main loop we introduce new indices (Val*, current and next), that are based on double buffering loop
  - through Double Buffering Pass we actively use and update data that has been created and store during allocation and memory aliasing pass (e.g. `LoadStoreOp` -> smem TV / `kir::Allocate` collection / `LoadStoreOp` -> loop index), it is later used by indexing pass

@naoyam, @zasdfgbnm - can I ask you to check if the direction I'm heading is aligned with Fuser current and future architecture? Thanks!

---

**NOTE**: for the time being, please ignore changes in `index_compute.cpp` - these need to be redone and refined.

---

The overview of how result cloned loops should look like:
```
// pre-prolog:
// - smem allocations (mbarriers, tokens)
// - mbarrier init (0..stages)
//
// prolog loop:
// - 0th thread:
//   - issue 0..stages-1 cp async bulks
//
// main loop:
// - 0th thread:
//   - issue next cp async bulk
// - copy body, without
//   - smem allocations
//   - mbarrier inits
//   - mbarrier inval
//
// epilogue loop:
// - copy body, without
//   - smem allocations
//   - issuing cp async
//   - mbarrier inits
//   - mbarrier inval
//
// post-epilogue:
//  - 0th thread: loop with mbarriers inval
```

The scope of kir changes in prolog loop:
```
// Replace cpAsyncBulk type LoadStoreOp with:
// if (0th thread in block)
//    token[loop_idx] =
//    mbarrier::arriveExpectTx(mbarrier[loop_idx])
//    cpAsyncBulk(mbarrier[loop_idx],...)
//
//  Where loop_idx is in range 0...stages-1
```

The scope of kir changes in main loop:
```
// Replace LoadStoreOp with:
//  if (0th thread in block)
//    token[next_stage] =
//    mbarrier::arriveExpectTx(mbarrier[next_stage])
//    cpAsyncBulk(mbarrier[next_stage],...)
//  mbarrier::wait(token[curr_stage])
//
// Where mbarrier and token are smem arrays bound to the LoadStoreOp
```

**Comment by naoyam** [comment]
> The state of the implementation, as of 02/22/2024:
> 
> * add handling of additional smem allocations needed to store tokens (token - an object used to synchronize memory operations that use mbarrier objects)
> * update smem allocations for mbarrier and token objects; double/circular buffering requires a number of stages of instances of these objects, per double-buffered TensorView
> * update mechanism for calculating smem allocations' lifetime (e.g. mbarrier starts before its initialize, and ends after invalidation)
> * modify indexing - we shift focus from allocation/index pass to double buffering pass w.r.t. cpAsyncBulk memory operations, but there are still cases where memory operation is cpAsyncBulk, but it is not a part of double buffering - for these cases we keep naïve synchronization mechanism, but at the same time we don't want to break what was done in Double Buffering pass (`LoadStoreOp` replacements / new loop introduction)
> * modify Double Buffering pass in lowering
>   
>   * separate execution paths, based on `LoadStoreOp`s in Double Buffer loop:
>     
>     * if there is no `LoadStoreOp` that is cpAsyncBulk, then use original (but with some changes) execution path (see `insert`)
>     * if there is at least one `LoadStoreOp` that is cpAsyncBulk, then use `insertTMA` (name to be changed) which redefines structure of result kir
>   * new execution path introduces two new concepts, apart form existing prolog/main/epilogue loops (more details can be found at the end of this long comment):
>     
>     * 'pre-prolog' - a collection of expressions that contain smem allocations (so these are visible through the double-buffered loop) and `mbarrier` initialization, which is predicated so it is done only by the first thread in a block,
>     * `post-epilogue` - a collection of expressions that handle `mbarrier` objects invalidation,
>   * to handle `cpAsyncBulk` / `mbarrier::wait` in main loop we introduce new indices (Val*, current and next), that are based on double buffering loop
>   * through Double Buffering Pass we actively use and update data that has been created and store during allocation and memory aliasing pass (e.g. `LoadStoreOp` -> smem TV / `kir::Allocate` collection / `LoadStoreOp` -> loop index), it is later used by indexing pass
> 
> @naoyam, @zasdfgbnm - can I ask you to check if the direction I'm heading is aligned with Fuser current and future architecture? Thanks!

It looks reasonable to me. At least, it seems like a natural extension of the current double buffering pass.

File: csrc/device_lower/pass/allocation.cpp [hunk 1]
```diff
@@ -551,25 +564,80 @@ class AllocationInserter : public kir::ExprMutator {
...
+                .contiguity(true)
+                .build();
+        mbarrier_tokens->setMemoryType(MemoryType::Shared);
+        kir::Allocate* mbarrier_tokens_alloc = IrBuilder::create<kir::Allocate>(
+            mbarrier_tokens, MemoryType::Shared);
+
+        registerInsertBefore(expr, mbarrier_tokens_alloc, expr_scope);
+        registerInsertBefore(expr, mbarrier_alloc, expr_scope);
+        registerInsertBefore(expr, mbarrier_init, expr_scope);
+        registerInsertAfter(expr, mbarrier_inval, expr_scope);
```
Comment (zasdfgbnm) [inline csrc/device_lower/pass/allocation.cpp:606]: So we are doing the init and inval inside the loop of cp.async, i.e. generate code like:
```
for i in range(I0):
    __shared__ tokens;
    __shared__ mbarrier;
    init(mbarrier);
    cp.async.bulk(data, mbarrier);
    inval(mbarrier);
```
and leave it for the double bufferring pass to fix it? Could you leave a comment in the code for this?

File: csrc/device_lower/pass/double_buffer.cpp [hunk 2]
```diff
@@ -419,6 +729,191 @@ class DoubleBufferLoopNestInspector : private kir::IrVisitor {
...
+//  and initialization of mbarriers.
+//
+// Expected result:
+//   mbarrier and token smem allocations
+//   if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
+//     for (unsigned i = 0; i < stages; ++i) {
+//       mbarrier::init(...);
+//     }
+//   }
+class CpAsyncBulkPrePrologue : public kir::IrVisitor {
```
Comment (zasdfgbnm) [inline csrc/device_lower/pass/double_buffer.cpp:743]: Should we add a `DoubleBufferLoopStage::PrePrologue` and a `DoubleBufferLoopStage::PostEpilogue` and move the code below to `DoubleBufferLoopCloner`?

File: csrc/device_lower/pass/allocation.cpp (hunk 1 above)
Comment (drzejan2) [reply csrc/device_lower/pass/allocation.cpp:606]: Yes, that's the current implementation - I will update the code here so it is clear that Double Buffering pass modify this part of the code. Having it done here has the benefit of easier control of shared memory allocations / storage liveness (the order of passes is `insertAllocations`-> `reuseMemoryAllocations` -> `DoubleBufferPass`).

Switching to approach, where all needed elements are added in double buffering pass (we detect `LoadStoreOp` for bulk copy op -> we add placeholder for tokes, mbarriers, init mbarrier and later inval for mbarriers) will have major consequences: adding to the system smem allocation (mbarrier / tokens) and life time analysis out-of-the order (pass-wise) might be a challenge.

I will think how we could have this done, in a clean way.

Thank you for the comment!

File: csrc/device_lower/pass/double_buffer.cpp (hunk 2 above)
Comment (drzejan2) [reply csrc/device_lower/pass/double_buffer.cpp:743]: That's a good point.

For now I have it as separate class of functional classes - the main reason is that `DoubleBufferLoopCloner` returns `kir::ForLoop` while pre-prologue/post-epilogue are a simple collection (a list) of expressions. But you are right, we could have this done in smarter way where these two new 'regions' are handled along prolog/main/epilogue. I will try to unify this part of code, or at least leave a comment that architecture of this part of system should be reconsidered.

Thanks!

File: csrc/device_lower/pass/allocation.cpp [hunk 3]
```diff
@@ -518,6 +520,17 @@ class AllocationInserter : public kir::ExprMutator {
...
+      // allocation
+      if (out_tv->isDoubleBuffered()) {
+        if (out_tv->isCircularBuffered()) {
+          circular_buffer_depth =
+              std::max(circular_buffer_depth, out_tv->circularBufferDepth());
+        } else {
+          circular_buffer_depth = std::max(circular_buffer_depth, 2u);
+        }
+      }
+
```
[the remark this answers is not in the document]
Comment (drzejan2) [reply csrc/device_lower/pass/allocation.cpp:533]: Now, that you have pointed this out, it really makes no sense to pick the largest stage. In fact, each circular buffer should be considered separately (and own set of mbarriers/tokens). I will update this part of code.

Thanks!

File: csrc/device_lower/pass/double_buffer.cpp [hunk 4]
```diff
@@ -289,11 +387,215 @@ class DoubleBufferLoopCloner : public kir::IrVisitor {
...
-         is_double_buffer_load_expr) ||
-        (loop_type_ == DoubleBufferLoopStage::Epilog &&
-         !is_double_buffer_load_expr)) {
-      cloned_scopes_.back()->push_back(expr);
+
+    if (has_cp_async_bulk_) {
+      auto gpu_lower = GpuLower::current();
+
+      // Expr that is a part of cpAsyncBulk synchronization proces, added
+      //  ealier to satisfy other passes checks, are handled already won't
```
Comment (rdspring1) [inline csrc/device_lower/pass/double_buffer.cpp:395]:
```suggestion
      //  earlier to satisfy other passes checks, are handled already won't
```

File: csrc/device_lower/pass/double_buffer.cpp [hunk 5]
```diff
@@ -289,11 +387,215 @@ class DoubleBufferLoopCloner : public kir::IrVisitor {
...
+      //  be pushed to the new scope.
+      // There can be cases where cpAsyncBulk data moevement is not a part of
```
Comment (rdspring1) [inline csrc/device_lower/pass/double_buffer.cpp:397]:
```suggestion
      // There can be cases where cpAsyncBulk data movement is not a part of
```

**Comment by rdspring1** [comment]
@drzejan2 I was reviewing this PR to see how it could be adapted for pointwise and normalization kernels. Is the PR in a usable state? What functionality is not yet implemented?

File: csrc/device_lower/pass/double_buffer.cpp (hunk 5 above)
Comment (drzejan2) [reply csrc/device_lower/pass/double_buffer.cpp]: Fixed along the rebase - thanks for pointing this out!

File: csrc/device_lower/pass/double_buffer.cpp (hunk 4 above)
Comment (drzejan2) [reply csrc/device_lower/pass/double_buffer.cpp]: Fixed along the rebase - thanks for pointing this out!

File: csrc/device_lower/pass/alias_memory.cpp [hunk 6]
```diff
@@ -869,17 +869,39 @@ class AllocationInfoMap : private kir::IrVisitor {
...
+        alloc_info->outer_live_interval->markWrite(write_pos);
+      };
+
+      markWrite(init->mbarrier()->as<TensorView>());
+
+      // Register life time start for a smem placeholder with tokens
+      //  returned by MBarrierArriveExpectTx / MBarrierArrive
+      if (GpuLower::current()->ldstMBarrierTokenMap().count(expr)) {
+        markWrite(GpuLower::current()->ldstMBarrierTokenMap()[expr]);
+      }
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply csrc/device_lower/pass/alias_memory.cpp:887]: The mbarrier and its token are mapped together. The token is the mbarrier state of the last phase.

Token liveness doesn't seem a strong priority. It is only a single `int64_t`.

**Comment by rdspring1** [comment]
Closed in lieu of https://github.com/NVIDIA/Fuser/pull/2833

**Pull request closed without merging**